File: 121.best-time-to-buy-and-sell-stock.py
#
# @lc app=leetcode id=121 lang=python3
#
# [121] Best Time to Buy and Sell Stock
#

# @lc code=start
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        profit = l = 0
        for r, cur in enumerate(prices[1:], start=1):
            if cur > prices[l]:
                profit = max(profit, cur - prices[l])
            else:
                l = r
        return profit


# @lc code=end

File: 121.best-time-to-buy-and-sell-stock.rs
/*
 * @lc app=leetcode id=121 lang=rust
 *
 * [121] Best Time to Buy and Sell Stock
 */

// @lc code=start
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        prices
            .into_iter()
            .fold((0, i32::MAX), |(profit, cost), price| {
                (profit.max(price - cost), price.min(cost))
            })
            .0
    }
}
// @lc code=end
